src/math/quat.zig

```diff
@@ -32,3 +280,212 @@ test "init" {
+    try testing.expect(f32, actual.v.w()).eqlApprox(expected.v.w(), 0.001);
+}
+
+test "conjugate" {
+    const q = math.Quat.init(1.0, 2.0, 3.0, 4.0);
+    const expected = math.Quat.init(-1.0, -2.0, -3.0, 4.0);
+    const actual = math.Quat.conjugate(&q);
+
+    try testing.expect(f32, actual.v.x()).eqlApprox(expected.v.x(), 0.001);
+    try testing.expect(f32, actual.v.y()).eqlApprox(expected.v.y(), 0.001);
+    try testing.expect(f32, actual.v.z()).eqlApprox(expected.v.z(), 0.001);
+    try testing.expect(f32, actual.v.w()).eqlApprox(expected.v.w(), 0.001);
+}
+
+test "fromMat4" {
+    const m = math.Mat4x4.rotateX(0.785398);
+    const q = math.Quat.fromMat(math.Mat4x4, &m);
+
+    try testing.expect(f32, q.v.x()).eqlApprox(0.383, 0.001);
+    try testing.expect(f32, q.v.y()).eqlApprox(0.0, 0.001);
+    try testing.expect(f32, q.v.z()).eqlApprox(0.0, 0.001);
+    try testing.expect(f32, q.v.w()).eqlApprox(0.924, 0.001);
+}
+
+test "fromEuler" {
+    const q = math.Quat.fromEuler(0.785398, 0.0, 0.0);
+
+    try testing.expect(f32, q.v.x()).eqlApprox(0.383, 0.001);
+    try testing.expect(f32, q.v.y()).eqlApprox(0.0, 0.001);
+    try testing.expect(f32, q.v.z()).eqlApprox(0.0, 0.001);
+    try testing.expect(f32, q.v.w()).eqlApprox(0.924, 0.001);
+}
+
+test "dot" {
+    const a = math.Quat.init(1.0, 2.0, 3.0, 4.0);
+    const b = math.Quat.init(5.0, 6.0, 7.0, 8.0);
+    const expected = 70.0;
+    const actual = math.Quat.dot(&a, &b);
+
+    try testing.expect(f32, actual).eqlApprox(expected, 0.001);
+}
+
+test "lerp" {
+    const a = math.Quat.init(1.0, 2.0, 3.0, 4.0);
+    const b = math.Quat.init(5.0, 6.0, 7.0, 8.0);
+    const expected = math.Quat.init(3.0, 4.0, 5.0, 6.0);
+    const actual = math.Quat.lerp(&a, &b, 0.5);
+
+    try testing.expect(f32, actual.v.x()).eqlApprox(expected.v.x(), 0.001);
+    try testing.expect(f32, actual.v.y()).eqlApprox(expected.v.y(), 0.001);
+    try testing.expect(f32, actual.v.z()).eqlApprox(expected.v.z(), 0.001);
+    try testing.expect(f32, actual.v.w()).eqlApprox(expected.v.w(), 0.001);
+}
+
+test "len2" {
+    const q = math.Quat.init(1.0, 2.0, 3.0, 4.0);
+    const expected = 30.0;
+    const actual = math.Quat.len2(&q);
+
+    try testing.expect(f32, actual).eqlApprox(expected, 0.001);
+}
+
+test "len" {
+    const q = math.Quat.init(0.0, 0.0, 3.0, 4.0);
+    const expected = 5.0;
+    const actual = math.Quat.len(&q);
+
+    try testing.expect(f32, actual).eqlApprox(expected, 0.001);
+}
+
+test "normalize" {
+    const q = math.Quat.init(0.0, 0.0, 3.0, 4.0);
+    const expected = math.Quat.init(0.0, 0.0, 0.6, 0.8);
+    const actual = math.Quat.normalize(&q);
+
+    try testing.expect(f32, actual.v.x()).eqlApprox(expected.v.x(), 0.001);
+    try testing.expect(f32, actual.v.y()).eqlApprox(expected.v.y(), 0.001);
+    try testing.expect(f32, actual.v.z()).eqlApprox(expected.v.z(), 0.001);
+    try testing.expect(f32, actual.v.w()).eqlApprox(expected.v.w(), 0.001);
```

Comment:
Can we make all these tests just use `eql`? It should be approximate equality with the epsilon already, and so should be good enough for testing in general